src/pharmpy/modeling/expressions.py

```diff
@@ -928,25 +949,36 @@ def get_pk_parameters(model: Model, kind: str = 'all') -> List[str]:
     get_individual_parameters
 
     """
+    natural_assignments = _get_natural_assignments(model.statements.before_odes)
+    cs_remapped = _remap_compartmental_system(model.statements, natural_assignments)
 
-    cs = model.statements.ode_system.to_compartmental_system()
+    free_symbols = set(_pk_free_symbols(cs_remapped, kind))
 
-    classified_assignments = list(
-        _classify_assignments(list(_assignments(model.statements.before_odes)))
-    )
+    dependency_graph = _dependency_graph(natural_assignments)
 
-    for t, assignment in reversed(classified_assignments):
-        if t == 'synthetic':
-            # NOTE Substitution must be made in this order
-            cs = cs.subs({assignment.symbol: assignment.expression})
+    return sorted(_filter_symbols(dependency_graph, free_symbols))
 
-    free_symbols = set(_pk_free_symbols(cs, kind))
 
-    assignments = list(_remove_synthetic_assignments(classified_assignments))
+def _get_natural_assignments(before_odes):
+    # Return assignments where assignments that are constants (e.g. X=1),
+    # single length expressions (e.g. S1=V), and divisions between parameters
+    # (e.g. K=CL/V) have been filtered out
+    classified_assignments = list(_classify_assignments(list(_assignments(before_odes))))
+    natural_assignments = list(_remove_synthetic_assignments(classified_assignments))
+    return natural_assignments
 
-    dependency_graph = _dependency_graph(assignments)
 
-    return sorted(_filter_symbols(dependency_graph, free_symbols))
+def _remap_compartmental_system(sset, natural_assignments):
+    # Return compartmental system where rates that are synthetic assignments
+    # have been substituted with their full definition (e.g K -> CL/V
+    cs = sset.ode_system.to_compartmental_system()
+
+    assignments = list(_assignments(sset.before_odes))
+    for assignment in reversed(assignments):
+        if assignment not in natural_assignments:
+            # NOTE Substitution must be made in this order
+            cs = cs.subs({assignment.symbol: assignment.expression})
+    return cs
```

Comment:
It is unfortunate that this now does containment tests on a `list` which is unnecessarily slow.

Reply:
That's true. Previously I had them in the same function, but felt it was a little messy to return the assignments and the subbed compartmental system from the same function (e.g. the subbed system is not needed for `get_rv_parameters`), and nice to have them separated (maybe clearer what they do when they are in two functions). The list of natural assignments should be relatively short since it's already filtered from the full assignments, not sure how slow this would be.

Reply:
You can always create a temporary `set` object if `Assignment` has a `__hash__` method.